httpio_async/__init__.py

```diff
@@ -0,0 +1,293 @@
+"""This module provides an interface for using httpio with the python3 asyncio library.
+
+The interface is where possible as similar to the existing httpio interface as possible (and hence similar to the
+file like objects of python) except that many methods are replaced with asynchronous coroutines."""
+
+import aiohttp
+from httpio import HTTPIOError
+
+
+__all__ = ["AsyncHTTPIOFile", "HTTPIOError", "open"]
+
+
+async def open(url, block_size=-1, **kwargs):
+    """
+    Open a URL as an asynchronous file-like object
+
+    The returned object is open, and needs to be closed when done.
+    This will be handled automatically if you use it as an asynchronous context manager.
+
+    :param url: The URL of the file to open
+    :param block_size: The cache block size, or `-1` to disable caching.
+    :param kwargs: Additional arguments to pass to `requests.Request()`
+    :return: An `httpio.HTTPIOFile` object supporting most of the usual
+        file-like object methods.
+    """
+    f = AsyncHTTPIOFile(session, url, block_size, **kwargs)
+    await f.open()
+    return f
+
+
+class AsyncHTTPIOFile(object):
+    """An asynchronous equivalent to httpio.HTTPIOFile.
+    Sadly this class cannot descend from that one for technical reasons.
+    """
+    def __init__(self, url, block_size=-1, **kwargs):
+        """
+        :param url: The URL of the file to open
+        :param block_size: The cache block size, or `-1` to disable caching.
+        :param kwargs: Additional arguments to pass to `session.get`
+        """
+        super(AsyncHTTPIOFile, self).__init__()
+        self.url = url
+        self.block_size = block_size
+
+        self._kwargs = kwargs
+        self._cursor = 0
+        self._cache = {}
+        self._session = None
+        self._aiter = None
+
+        self.length = None
+        self.closed = True
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self):
+        if self._aiter is None:
+            self._aiter = self.readlines()
+
+        return await self._aiter.__anext__()
+
+    async def open(self):
+        """This method has no direct equivalent in the HTTPIOFile interface, but does things which
+        are the equivalent of what's done in the constructor of that class. Since constructors cannot
+        be coroutines this class needs this as a seperate coroutine"""
+
+        if self.closed:
```

Comment:
Should this check `_session` instead to be consistent with SyncHTTPIO, or SyncHTTPIO be changed? E.g. if `open` fails with an exception raised then a second attempt should do nothing to be consistent. I haven't checked when BufferedIOBase or IOBase set `closed` and that might be relevant here as well.

Reply:
The base classes set closed to False upon initialisation, which is different behaviour from the Async version.